detector/standalone.go

```diff
@@ -1,48 +1,199 @@
-	// 	}
-	// }()
+	log.V(2).Info("Detect()")
+	s.pollOnce.Do(func() {
+		log.V(1).Info("spinning up asyc master detector poller")
+		go s.poller()
+	})
+	if o != nil {
+		log.V(1).Info("spawning asyc master detector listener")
+		go func() {
+			log.V(2).Infof("waiting for polled to send updates")
+		pollWaiter:
+			for {
+				select {
+				case mi, ok := <-s.ch:
+					if !ok {
+						break pollWaiter
+					}
+					log.V(1).Infof("detected master change: %+v", mi)
+					o.Notify(mi)
+				case <-s.done:
+					return
+				}
+			}
+			o.Notify(nil)
+		}()
+	} else {
+		log.Warningf("detect called with a nil master change listener")
+	}
 	return nil
 }
 
-// Stop the detection.
-func (s *Standalone) Stop() {
-	select {
-	case <-s.ch:
-		// already closed, don't close it again
-	default:
-		close(s.ch)
+func (s *Standalone) Done() <-chan struct{} {
+	return s.done
+}
+
+func (s *Standalone) Cancel() {
+	s.cancelOnce.Do(func() { close(s.done) })
+}
+
+// poll for changes to master leadership
+func (s *Standalone) poller() {
+	if s.initial == nil {
+		log.Warningf("aborting master poller since initial master info is nil")
+		return
+	}
+	//TODO(jdef) we could attempt to unpack IP addres if Host=""
+	if s.initial.Hostname != nil && len(*s.initial.Hostname) == 0 {
+		log.Warningf("aborted mater poller since initial master info has no host")
+		return
+	}
+	addr := *s.initial.Hostname
+	port := uint32(defaultMesosMasterPort)
+	if s.initial.Port != nil && *s.initial.Port != 0 {
+		port = *s.initial.Port
+	}
+	addr = net.JoinHostPort(addr, strconv.Itoa(int(port)))
+	log.V(1).Infof("polling for master leadership at '%v'", addr)
+	var lastpid *upid.UPID
+	for {
+		startedAt := time.Now()
+		ctx, cancel := context.WithTimeout(context.Background(), mesosLeaderSyncInterval)
+		if pid, err := s.fetchPid(ctx, addr); err == nil {
+			if !pid.Equal(lastpid) {
+				log.V(2).Infof("detected leadership change from '%v' to '%v'", lastpid, pid)
+				lastpid = pid
+				elapsed := time.Now().Sub(startedAt)
+				mi := CreateMasterInfo(pid)
+				select {
+				case s.ch <- mi:
+					// noop
+				case <-time.After(mesosLeaderSyncInterval - elapsed):
+					// no one heard the master change, oh well - poll again
+					cancel()
+					continue
+				case <-s.done:
+					cancel()
+					return
+				}
+			} else {
+				log.V(2).Infof("no change to master leadership: '%v'", lastpid)
+			}
+		} else {
+			log.Warning(err)
+		}
+		remaining := mesosLeaderSyncInterval - time.Now().Sub(startedAt)
+		log.V(3).Infof("master leader poller sleeping for %v", remaining)
+		time.Sleep(remaining)
+		cancel()
 	}
 }
 
-func (s *Standalone) Done() <-chan struct{} {
-	return nil
+func (s *Standalone) fetchPid(ctx context.Context, address string) (*upid.UPID, error) {
+	//TODO(jdef) need better address parsing, for now assume host, or host:port format
+	//TODO(jdef) need SSL support
+	uri := fmt.Sprintf("http://%s/state.json", address)
+	req, err := http.NewRequest("GET", uri, nil)
```

Comment:
So if specified master is dead, no way to find out new leader, correct ?  It be sweet (probably not necessary) to be able to pass in a list of known masters.  If req to get leader fails, try the next master addr.  Looks ok now.